aws-android-sdk-s3/src/test/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadInputStreamTest.java

```diff
@@ -0,0 +1,221 @@
+                eq(metadata), isNull(CannedAccessControlList.class), isNull(TransferListener.class));
+    }
+
+    @Test
+    public void testUploadWithBucketAndCannedAcl() throws IOException {
+        String key = getRandomString();
+        String bucket = getRandomString();
+        transferUtility.upload(key, inputStream,
+                UploadOptions.builder().bucket(bucket).cannedAcl(cannedAcl).build());
+        verify(transferUtility, times(1)).upload(eq(bucket), eq(key), any(File.class),
+                any(ObjectMetadata.class), eq(cannedAcl), isNull(TransferListener.class));
+    }
+
+    @Test
+    public void testUploadWithBucketAndMetadataAndCannedAcl() throws IOException {
+        String key = getRandomString();
+        String bucket = getRandomString();
+        transferUtility.upload(key, inputStream,
+                UploadOptions.builder().bucket(bucket).objectMetadata(metadata).cannedAcl(cannedAcl).build());
+        verify(transferUtility, times(1)).upload(eq(bucket), eq(key), any(File.class),
+                eq(metadata), eq(cannedAcl), isNull(TransferListener.class));
+    }
+
+    @Test
+    public void testUploadWithMetadataAndCannedAclAndListener() throws IOException {
+        String key = getRandomString();
+        transferUtility.upload(key, inputStream,
+                UploadOptions.builder().objectMetadata(metadata).cannedAcl(cannedAcl).transferListener(listener).build());
+        verify(transferUtility, times(1)).upload(eq(DEFAULT_BUCKET), eq(key),
+                any(File.class), eq(metadata), eq(cannedAcl), eq(listener));
+    }
+
+    @Test
+    public void testUploadWithAllParameters() throws IOException {
+        String key = getRandomString();
+        String bucket = getRandomString();
+        transferUtility.upload(key, inputStream,
+                UploadOptions.builder()
+                             .bucket(bucket)
+                             .objectMetadata(metadata)
+                             .cannedAcl(cannedAcl)
+                             .transferListener(listener)
+                             .build());
+        verify(transferUtility, times(1)).upload(eq(bucket), eq(key), any(File.class),
+                eq(metadata), eq(cannedAcl), eq(listener));
+    }
+
+    // Verify that the File does in fact get deleted after the upload is complete.
+    @Test
+    public void testDeleteTempFileAfterUpload() throws IOException {
+        String key = getRandomString();
+        final TransferObserver observer = transferUtility.upload(key, inputStream);
+        observer.setTransferListener(new TransferListener() {
+            @Override
+            public void onStateChanged(int id, TransferState state) {
+                Log.d(TAG, "onStateChanged: " + id + ", " + state);
+                switch (state) {
+                    case COMPLETED:
+                        assertFalse(new File(observer.getAbsoluteFilePath()).exists());
```

Comment:
As written, I believe this test will succeed before `onStateChanged` is called, so this assert statement will never be evaluated.   

// Define a `CountdownLatch` at the top of the test
```
CountDownLatch deleteFileLatch = new CountDownLatch(1);
```

In the callback, call countDown if the file has been deleted:
```
case COMPLETED:
    if(!new File(observer.getAbsoluteFilePath()).exists()) {
        deleteFileLatch.countDown();
    }
```

Then, at the end of the test, validate that the latch counts down within some time period, like 5 seconds:
```
if (!deleteFileLatch.await(2, TimeUnit.SECONDS)) {
    fail("File not deleted after upload.");
}
```